wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DockerimageToolApplier.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static com.google.common.base.Strings.*;
+import static org.eclipse.che.api.core.model.workspace.config.Command.MACHINE_NAME_ATTRIBUTE;
+import static org.eclipse.che.api.core.model.workspace.config.MachineConfig.MEMORY_LIMIT_ATTRIBUTE;
+import static org.eclipse.che.api.devfile.server.Constants.DOCKERIMAGE_TOOL_TYPE;
+import static org.eclipse.che.api.workspace.shared.Constants.PROJECTS_VOLUME_NAME;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Endpoint;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.MachineConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+import org.eclipse.che.api.workspace.server.model.impl.ServerConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.VolumeImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.KubernetesSize;
+
+/**
+ * Applies Kubernetes tool configuration on provided {@link Devfile} and {@link
+ * WorkspaceConfigImpl}.
+ *
+ * @author Sergii Leshchenko
+ */
+public class DockerimageToolApplier {
+
+  private static final String DOCKERIMAGE_RECIPE_TYPE = "dockerimage";
+
+  private final String projectFolderPath;
+
+  @Inject
+  public DockerimageToolApplier(@Named("che.workspace.projects.storage") String projectFolderPath) {
+    this.projectFolderPath = projectFolderPath;
+  }
+
+  /**
+   * Applies Kubernetes tool configuration on provided {@link Devfile} and {@link
+   * WorkspaceConfigImpl}.
+   *
+   * <p>It includes:
+   *
+   * <ul>
+   *   <li>provisioning environment based on specified configuration in tool: dockerimage, env,
+   *       volumes, etc.;
+   *   <li>provisioning machine name attribute to commands that are configured to be run in the
+   *       specified tool;
+   * </ul>
+   *
+   * @param tool dockerimage tool that should be applied to devfile and workspace config
+   * @param devfile devfile that should be changed according to the provided tool
+   * @param workspaceConfig workspace config that should be changed according to the provided tool
+   * @throws IllegalArgumentException when wrong type tool is passed
+   */
+  public void apply(Tool tool, Devfile devfile, WorkspaceConfigImpl workspaceConfig) {
```

Comment:
To be honest, for me this method signature looks a bit strange: we're passing tool, and then whole devfile containing this tool.  
Since we're just modifying the commands from a devfile, maybe we just can pass commands list? For me it it will look clearer a bit and javadoc can be more specific than (same can be applied to the K8S tool applier).  WDYT ?

Reply:
Well, I don't think that putting the command list as an argument is a good way to refactor it.
`ToolApplier` is an abstraction that may modify any part of `devfile` and now it's only command list.
As alternative I think we can refactor it in the following way:
```
interface ToolApplier {
    void apply(Devfile devfile, WorkspaceConfigImpl workspaceConfig);
}

class EditorToolApplier ...
class PluginToolApplier ... //
class KubernetesToolApplier ... // handle openshift as well
class DockerimageToolApplier ...
```
Then `DevfileConverter` won't contain any tool implementation specific logic and will only invoke `ToolApplier` implementation that may be bind as `Set`
@mshaposhnik WDYT?

Reply:
Yep looks good. This will allow to concentrate logic in single place instead of spreading it.

Reply:
I've already done some part of the needed refactoring but it would be easier to review dockerimage support and refactoring separately. So, a new PR will be created after this PR is merged.

Reply:
A separate issue is created to do this refactoring https://github.com/eclipse/che/issues/12659